### PR TITLE
Implement push notifications

### PR DESCRIPTION
**Pull request by BentiGorlich**
- add a periodic fetch for the notifications count
- add a push notification system -> each browser generates a device key with which a push subscription is associated
- a bit of prep work for handling push notifications for apps, though that needs more work
- each notification can generate a `PushNotification` object with a title, text and some URLs for the push notifications
- creating a `NotificationCreatedEvent` to send the push subscriptions

Since push notifications need some encryption you gotta run this command once the code is running:

```sh
php bin/console mbin:push:keys:update
```

@jwr1 I will need your help to implement push notifications for Interstellar / 3rd party apps

### CODE REVIEW

File: public/sw.js
Comment (melroy89) [inline public/sw.js:2]: found it!.. haha off-topic

**Comment by BentiGorlich** [comment]
I have it running on my server now and it seems like I have some problems on mobile, though that is a topic for another day

**Comment by BentiGorlich** [comment]
Fixed the mobile problems I had :)

**Comment by melroy89** [comment]
I'm testing it now at kbin.melroy.org atm

**Comment by melroy89** [comment]
Can we hide the "Register for push notification" when you approved it? Aka, you can you read the status back from the browser whether or not the user approved push notifications?

**Comment by BentiGorlich** [comment]
That is exactly what it is doing, however if there was a problem on the server it will delete the push subscription again. Did you execute this command: `php bin/console mbin:push:keys:update`? If yoh did, check the logs for errors

**Comment by melroy89** [comment]
> That is exactly what it is doing, however if there was a problem on the server it will delete the push subscription again. Did you execute this command: `php bin/console mbin:push:keys:update`? If yoh did, check the logs for errors

AH yes! My bad.. Forgot to run that, despite you explicitly mentioned it. 

Test push notification also works. This is great!

**Review by melroy89 — APPROVED** [review]

**Comment by melroy89** [comment]
Ready to merge or?

**Comment by BentiGorlich** [comment]
I gotta add the API endpoint for registering for push notifications, but maybe just do it in a new PR?

**Comment by BentiGorlich** [comment]
Since I will probably have to adjust the migration I want to implement it in this PR

**Comment by melroy89** [comment]
> Since I will probably have to adjust the migration I want to implement it in this PR

Sue I understand. Hopefully John can help you there.

**Comment by BentiGorlich** [comment]
For those who already ran this PR on their server and didn't revert the migration: execute the following:
```sql
ALTER TABLE oauth2_user_consent DROP CONSTRAINT fk_c8f05d01562830f3;
DROP INDEX uniq_c8f05d01562830f3;
ALTER TABLE oauth2_user_consent DROP push_subscription;
ALTER TABLE user_push_subscription DROP CONSTRAINT fk_ae378bd83b1f161a;
DROP INDEX uniq_ae378bd83b1f161a;
ALTER TABLE user_push_subscription DROP user_consent;
ALTER TABLE user_push_subscription ADD api_token CHAR(80) DEFAULT NULL;
ALTER TABLE user_push_subscription ADD CONSTRAINT FK_AE378BD87BA2F5EB FOREIGN KEY (api_token) REFERENCES oauth2_access_token (identifier) ON DELETE CASCADE NOT DEFERRABLE INITIALLY IMMEDIATE;
CREATE UNIQUE INDEX UNIQ_AE378BD87BA2F5EB ON user_push_subscription (api_token);
```

But I'd suggest just reverting the migration beforehand :)

**Comment by BentiGorlich** [comment]
Ok, so I added an API endpoint @jwr1 . I might very well have overlooked something or it straight up does not work, we'll see. I think this PR can be merged, even if the API is not working correctly. The DB will probably not be changed again

**Comment by nobodyatroot** [comment]
Coolio! Do you know why Android might not be properly showing the Mbin icon in the taskbar? GS24U, Android 14, and Chrome

image

**Comment by nobodyatroot** [comment]
> Coolio! Do you know why Android might not be properly showing the Mbin icon in the taskbar? GS24U, Android 14, and Chrome
> 
> 
> The weird thing is the pulldown then shows the Mbin PWA icon:
> 

ok, i think i answered my own question

image

https://github.com/MbinOrg/mbin/pull/921/files#diff-98fd1d4c44ad3934ee242b1f070d235cc39f11e46738f28de66c6740182c4706R11

I don't see `badge` getting passed along... and apparently it needs to be a transparent all white (alpha channel) png icon for it to work correctly on Android.

**Comment by BentiGorlich** [comment]
No that is not it: 
```js
self.registration.showNotification(json.title, { body: json.message, data: json, icon: json.avatarUrl ?? json.iconUrl, badge: json.iconUrl })
```
I do pass the icon as the `badge`. Though on FF for Android I only see the Firefox icon as well...

**Comment by BentiGorlich** [comment]
I did not test Chrome though, so it might be worth testing more stuff

**Review by nobodyatroot — APPROVED** [review]
I guess I don't want to hold this up over icon behavior, especially since we'll have to check a ton of platforms.... maybe a better option to fix in a future PR.

**Comment by BentiGorlich** [comment]
I pushed a new icon which does not have a background. @nobodyatroot can you check how chrome displays it now?

**Comment by nobodyatroot** [comment]
> I pushed a new icon which does not have a background. @nobodyatroot can you check how chrome displays it now?

making progress! the notification bar now shows a white icon, but the pulldown is still kinda messed up.

image

image

can you just export the alpha channel? the raw asset still has color, so that may be causing issues.

**Review by nobodyatroot — APPROVED** [review]
Icon issues (at least for Chrome users) is resolved! Looks good!

image

image

**Pull request merged**
